modules/core/src/smithy4s/Blob.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+
+import java.nio.ByteBuffer
+import java.util.Base64
+import java.nio.charset.StandardCharsets
+
+sealed trait Blob {
+  def toArray: Array[Byte]
+  def isEmpty: Boolean
+  final def toArrayBlob: Blob = Blob(toArray)
+  final def toBase64String: String = Base64.getEncoder().encodeToString(toArray)
+  final def toUTF8String: String = new String(toArray, StandardCharsets.UTF_8)
+}
+
+object Blob {
+
+  val empty: Blob = new Blob.ByteArrayBlob(Array.emptyByteArray)
+
+  def apply(bytes: Array[Byte]): Blob = new ByteArrayBlob(bytes)
+  def apply(buffer: ByteBuffer): Blob = new ByteBufferBlob(buffer)
+  def apply(string: String): Blob = new ByteArrayBlob(string.getBytes())
```

Comment:
thought: UTF-8 is only a standard encoding since Java 18, previously it was platform-dependent. Shall we hardcode UTF-8 while we support older JVMs?

Reply:
another thought, shall we make another variant of `Blob` that's optimized for strings?

Reply:
I thought about it making a string specialised case and decided against it as it'd make the integrations more annoying.

I'm happy to hardcore utf-8 in the smart constructors though.